elasticdl/python/master/pod_manager.py

```diff
@@ -14,7 +14,10 @@
 
 class PodManager(object):
     def __init__(self, args):
-        pass
+        self._observers = []
 
     def start(self):
         pass
+
+    def add_observer(self, observer):
+        self._observers.append(observer)
```

Comment:
In this way, when an observer is added, it requires to have 4 event handling functions (though some may be original empty implementation). For an event such as pod started, all observers are called.
The other way to implement is to call only those meaningful event handing functions. Then we need to add event handlers, not observers with 4 event handlers.

Reply:
Can you share an example about event handler way? It would be helpful for the comparison.

Reply:
May we refer to the implement  of `tf.keras.callbacks.Callback`.

Reply:
Yes, current design is similar to the keras callback

Reply:
Something like this:

```
class PodEventObserver(object):
  def __init__(self):
      callbacks = {"started": [], "failed": [], "succeeded": [], "deleted": []}

  def register_callback(self, event_type, callback):
    self.callbacks[event_type].append(callback)

  def make_callbacks(self, event_type, pod_info, cluster_context):
     for callback in self.callbacks[event_type]:
         callback(pod_info, cluster_context)

  def on_pod_started(self, pod_info, cluster_context):
     self.make_callbacks("started", pod_info, cluster_context)

   def on_pod_failed(self, pod_info, cluster_context):
     self.make_callbacks("failed", pod_info, cluster_context)

  ...

```

Reply:
```
class PodEventSubscriber(object):
  def __init__(self):
      callbacks = {"started": [], "failed": [], "succeeded": [], "deleted": []}

  def register_callback(self, event_type, callback):
    self.callbacks[event_type].append(callback)

  def make_callbacks(self, event_type, pod_info, cluster_context):
     for callback in self.callbacks[event_type]:
         callback(pod_info, cluster_context)
 ```

```
class PodEventObserver(object):
  def __init__(self):
      callbacks = {"started": [], "failed": [], "succeeded": [], "deleted": []}

  def on_pod_started(self, pod_info, cluster_context):
     self.make_callbacks("started", pod_info, cluster_context)

   def on_pod_failed(self, pod_info, cluster_context):
     self.make_callbacks("failed", pod_info, cluster_context)

  ...

Reply:
> ```
> class PodEventSubscriber(object):
>   def __init__(self):
>       callbacks = {"started": [], "failed": [], "succeeded": [], "deleted": []}
> 
>   def register_callback(self, event_type, callback):
>     self.callbacks[event_type].append(callback)
> 
>   def make_callbacks(self, event_type, pod_info, cluster_context):
>      for callback in self.callbacks[event_type]:
>          callback(pod_info, cluster_context)
> ```
> 
> ```
> class PodEventObserver(object):
>   def __init__(self):
>       callbacks = {"started": [], "failed": [], "succeeded": [], "deleted": []}
> 
>   def on_pod_started(self, pod_info, cluster_context):
>      self.make_callbacks("started", pod_info, cluster_context)
> 
>    def on_pod_failed(self, pod_info, cluster_context):
>      self.make_callbacks("failed", pod_info, cluster_context)
> 
>   ...
> ```

Reply:
class TaskObserver(PodEventObserver):
    def __init__(self, task_manager):
         self._task_manager = task_manager

   def on_pod_failed():
       self._task_manager.reassign_task()